### ISSUE TITLE
Better support for scalar model parameters

### ISSUE THREAD

**Issue by MikeInnes** [FluxML/Flux.jl#214]
`TrackedReal`s are still not first-class in that we don't support passing them to optimisers, for example. So we don't really support scalar model parameters, which e.g. you might use in a basic linear regression. It's easy to make this work with singleton or 0D arrays, but we should just support the obvious thing.

As an immediate step, `Param`s will need to be mutable and be able to support immutable param values.

**Comment by ssfrr** [comment]
Additionally, `Complex`-valued scalar values can't be constructed:

```julia
param(1+0im)
```
gives
```
MethodError: Cannot `convert` an object of type Float64 to an object of type Flux.Tracker.TrackedReal
This may have arisen from a call to the constructor Flux.Tracker.TrackedReal(...),
since type constructors fall back to convert methods.
convert at complex.jl:31 [inlined]
Flux.Tracker.TrackedReal(::Complex{Float64}) at sysimg.jl:77
param(::Complex{Int64}) at Tracker.jl:54
include_string(::String, ::String) at loading.jl:522
...
```

**Comment by MikeInnes** [comment]
This would require a new `TrackedComplex` type, but I think it'd be pretty easy to copy the Real implementation from [this file](https://github.com/FluxML/Flux.jl/blob/4320738d870bfefa07f99cbabdf9367f9e9bdb6f/src/tracker/scalar.jl) to get something going.

**Comment by kota7** [comment]
I would like the feature of supporting scalar parameters.

I observe that some optimisers store internal information as `IdDict`.  This would work with parameter as key (since parameters are mutable) but not necessarily with parameter data if scalar.

```
x = param(1.0)
d = IdDict()
d[x.data] = 100

@show haskey(d, x.data)  # true
update!(x, 1.0)
@show haskey(d, x.data)  # false
```

```
x = param(1.0)
d = IdDict()
d[x] = 100

@show haskey(d, x)  # true
update!(x, 1.0)
@show haskey(d, x)  # true
```

Whereas it seems optimisers are "applied" often with parameter values, instead of parameter itself.

https://github.com/FluxML/Flux.jl/blob/8386a49bf9a2950d51889b8af1ace178bf5169fa/src/optimise/train.jl#L9-L15

How about switching to `apply!(opt, x, x.grad)`?

**Comment by MikeInnes** [comment]
The IdDict is ok for now since scalar params are actually `TrackedReal` objects. It will be a problem in future, but that will be solved by the optimiser redesign in #637.
